driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/ClientSessionBinding.java

```diff
@@ -107,28 +84,49 @@ public OperationContext getOperationContext() {
         return wrapped.getOperationContext();
     }
 
-    private void getPinnedConnectionSource(final boolean isRead, final SingleResultCallback<AsyncConnectionSource> callback) {
+    @Override
+    public void getReadConnectionSource(final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrapped::getReadConnectionSource, callback);
+    }
+
+    @Override
+    public void getReadConnectionSource(final int minWireVersion, final ReadPreference fallbackReadPreference,
+            final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrappedConnectionSourceCallback ->
+                wrapped.getReadConnectionSource(minWireVersion, fallbackReadPreference, wrappedConnectionSourceCallback),
+                callback);
+    }
+
+    public void getWriteConnectionSource(final SingleResultCallback<AsyncConnectionSource> callback) {
+        getConnectionSource(wrapped::getWriteConnectionSource, callback);
+    }
+
+    private interface AsyncConnectionSourceSupplier {
+        void getConnectionSource(SingleResultCallback<AsyncConnectionSource> callback);
+    }
+
+    private void getConnectionSource(final AsyncConnectionSourceSupplier connectionSourceSupplier,
+            final SingleResultCallback<AsyncConnectionSource> callback) {
         WrappingCallback wrappingCallback = new WrappingCallback(callback);
-        TransactionContext<AsyncConnection> transactionContext = TransactionContext.get(session);
-        if (transactionContext == null) {
-            SingleResultCallback<AsyncConnectionSource> connectionSourceCallback = (result, t) -> {
+
+        if (!session.hasActiveTransaction()) {
+            connectionSourceSupplier.getConnectionSource(wrappingCallback);
+        } else if (TransactionContext.get(session) == null) {
```

Comment:
If we do the following, the control flow statements stay more similar to what we have in the synchronous code.

```suggestion
        if (!session.hasActiveTransaction()) {
            connectionSourceSupplier.getConnectionSource(wrappingCallback);
            return;
        }
        if (TransactionContext.get(session) == null) {
```

Reply:
I didn't like the `return` statements that this requires, but I see your point.